java/src/main/native/src/row_conversion.cu

```diff
@@ -0,0 +1,365 @@
+
+        for (cudf::size_type shared_index = shared_input_index; 
+                shared_index < shared_input_end;
+                shared_index += shared_input_stride) {
+            long_output[start_output_index + shared_index] = long_shared[shared_index];
+        }
+        __syncthreads();
+        // Go for the next round
+    }
+}
+
+static bool is_fixed_width(const cudf::column_view & v) {
+    return cudf::is_fixed_width(v.type());
+}
+
+static inline int32_t align_offset(int32_t offset, std::size_t alignment) {
+    return (offset + alignment - 1) & ~(alignment - 1);
+}
+
+static std::unique_ptr<cudf::column> fixed_width_convert_to_rows(
+        const cudf::size_type start_row,
+        const cudf::size_type num_rows,
+        const cudf::size_type num_columns,
+        const cudf::size_type size_per_row,
+        std::unique_ptr<rmm::device_uvector<cudf::size_type>> & column_start,
+        std::unique_ptr<rmm::device_uvector<cudf::size_type>> & column_size,
+        std::unique_ptr<rmm::device_uvector<const int8_t *>> & input_data,
+        std::unique_ptr<rmm::device_uvector<const cudf::bitmask_type *>> & input_nm,
+        const cudf::scalar & zero,
+        const cudf::scalar & scalar_size_per_row,
+        cudaStream_t stream,
+        rmm::mr::device_memory_resource* mr
+        ) {
+    int64_t total_allocation = size_per_row * num_rows;
+    if (total_allocation > std::numeric_limits<int>::max()) {
+        // We made a mistake in the split somehow
+        CUDF_EXPECTS(false, "Table is too large to fit");
+    }
+
+    // Allocate and set the offsets row for the byte array
+    std::unique_ptr<cudf::column> offsets = cudf::detail::sequence(
+            num_rows + 1, zero, scalar_size_per_row);
+
+    std::unique_ptr<cudf::column> data =
+        cudf::make_numeric_column(
+                cudf::data_type(cudf::type_id::INT8),
+                static_cast<cudf::size_type>(total_allocation));
+
+    // We have found speed degrades when a thread handles more than 4 columns.
+    // Each block is 2 dimensional and the y dimension indicates the number of
+    // columns.  We limit this to 32 threads in the y dimension so we can still
+    // have 32 threads in the x dimension which should result in coalesceing
+    // memory reads
+    int y_block_size = (num_columns + 3) / 4;
+    if (y_block_size > 32) {
+      y_block_size = 32;
+    }
+    int x_possible_block_size = 1024/y_block_size;
+    int max_shared_size = 48 * 1024; // 48KB is the default setting for shared memory per block according to the cuda tutorials
+    int max_block_size = max_shared_size/size_per_row;
+    // If we don't have enough shared memory there is no point in haveing more threads per block that will just sit idle
+    max_block_size = max_block_size > x_possible_block_size ? x_possible_block_size : max_block_size;
+    int block_size = (max_block_size / 32) * 32;
+    if (block_size == 0) {
```

Comment:
`CUDF_EXPECTS(block_size != 0, "Row size is too large to fit in shared memory");`